### PR TITLE
Issue #2671810 by Boobaa: Implement configuration form for View widget

### PR DESCRIPTION
**Pull request by boobaa**
[PR#142](https://github.com/drupal-media/entity_browser/pull/142) is related (as described at [the accompanying d.o issue](https://www.drupal.org/node/2366315#comment-10910876)).

### CODE REVIEW

File: src/Plugin/EntityBrowser/Widget/View.php [hunk 1]
```diff
@@ -209,4 +210,41 @@ public function submit(array &$element, array &$form, FormStateInterface $form_s
...
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $options = [];
+    // Get only those enabled Views that have entity_browser displays.
+    $displays = Views::getApplicableViews('entity_browser_display');
+    foreach ($displays as $display) {
+      list($view_id, $display_id) = $display;
+      $view = $this->entityManager->getStorage('view')->load($view_id);
+      $options[$view_id . '.' . $display_id] = $this->t('@view : @display', array('@view' => $view->label(), '@display' => $view->get('display')[$display_id]['display_title']));
+    }
```
Comment (dawehner) [inline src/Plugin/EntityBrowser/Widget/View.php:225]: IMHO it would be nice to tell people that they have to create views with the entity browser display, when they haven't done any yet.

**Comment by dawehner** [comment]
In general this looks alright. Do we need some kind of test coverage for it?

File: src/Plugin/EntityBrowser/Widget/View.php (hunk 1 above)
Comment (boobaa) [inline src/Plugin/EntityBrowser/Widget/View.php:225]: Well, I think we should think about two things.
1. Should we introduce a dependency on views.module in entity_browser.info.yml? Currently we don't have this, and we don't even check if Views is enabled, so one could enable Entity Browser without Views, and will end up having nothing but a cryptic error message saying entity type 'view' is not found, if at all. (This error message was already present before this PR.)
2. If we decide not to introduce a hard dependency on views.module (see above), then what should be the most appropriate place to check for it?

In other words: is there any real use case for Entity Browser without Views?

Comment (slashrsm) [inline src/Plugin/EntityBrowser/Widget/View.php:225]: > In other words: is there any real use case for Entity Browser without Views?

There are many. Views should not be a hard dependency of Entity browser. We should check if Views are enabled and warn users if they're trying to use views widget without it.

File: src/Plugin/EntityBrowser/Widget/View.php [hunk 2]
```diff
@@ -209,4 +210,41 @@ public function submit(array &$element, array &$form, FormStateInterface $form_s
...
+      '#options' => $options,
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state) {
```
Comment (slashrsm) [inline src/Plugin/EntityBrowser/Widget/View.php:241]: Submit should be handled in \Drupal\entity_browser\PluginConfigurationFormTrait. Individual plugins should have a need to implement their own submit handlers in rare cases.

Comment (boobaa) [inline src/Plugin/EntityBrowser/Widget/View.php:241]: I think this is such a case. The sitebuilder has to select a single view display for the widget, but this is stored as a (view, view_display) pair. I think it's easier to use a single dropdown for this (as the UI), but I didn't want to change how this config is stored (so I didn't have to change even more things). That's why I added a submitConfigurationForm method that is specific to this class: it stores one single input as two different strings.

(I was thinking about two dropdowns first: one for selecting the view, the other selecting a display of that view – but this approach would yield even more changes like an AJAX call since the display dropdown's content depends on the current selection in the view dropdown, etc. So I've taken the easier road here.)

File: src/Plugin/EntityBrowser/Widget/View.php (hunk 1 above)
Comment (boobaa) [inline src/Plugin/EntityBrowser/Widget/View.php:225]: How, when and where should we check that? Y'know, if Views is not a hard dependency, then somebody might set up an entity_browser with a view widget, then uninstall Views – and there might be even more shady cases hidden somewhere. Should we check for Views' availability every time it might be needed? Or… should we add a new property to \Drupal\entity_browser\Annotation\EntityBrowserWidget as an array of dependencies, so every widget could inform Entity Browser about its own dependencies? (Tho I don't yet know how to add that property, nor how, when and where should Entity Browser check a plugin's dependencies.)

And one more question: should this all be handled in this issue/PR, or should we just commit this and handle the dependencies in a different issue?

Comment (slashrsm) [inline src/Plugin/EntityBrowser/Widget/View.php:225]: There are many ways we could approach this problem. I'd say that we keep this simple and simply check if Views are enabled when displaying entity browser and display a warning. Additional to that we could do the same check/warning on the configuration UI page.

> And one more question: should this all be handled in this issue/PR, or should we just commit this and handle the dependencies in a different issue?

Separate issue/PR would work.

File: src/Plugin/EntityBrowser/Widget/View.php (hunk 2 above)
Comment (slashrsm) [inline src/Plugin/EntityBrowser/Widget/View.php:241]: OK, makes sense. Thank you for the clarification.

**Comment by ayalon** [comment]
Hi

I tested this pull request. The dropdown shows up and I'm able to select a view. But the view cannot be displayed in a modal dialog.

If I link a entity reference field with the entity browser (Modal) I get an error while opening the modal dialog:
Fatal error: Call to a member function getExecutable() on a non-object in 
../modules/contrib/entity_browser/src/Plugin/EntityBrowser/SelectionDisplay/View.php on line 47

I debugged this and found out, that the the configuration of the view and view_display is missing:

`
$this->configuration['view']
$this->configuration['view_display']
`
This variables are NULL and so the view cannot be found.

File: src/Plugin/EntityBrowser/Widget/View.php (hunk 1 above)
Comment (dawehner) [inline src/Plugin/EntityBrowser/Widget/View.php:225]: You could actually set `"provider" = "views"` in the annotation and `\Drupal\Core\Plugin\DefaultPluginManager::findDefinitions` will filter it out by default.

**Comment by dawehner** [comment]
@ayalon Yeah, well but this is about a different file. This PR is just about the widget

**Comment by ayalon** [comment]
But how can I access the configuration of the Widget from SelectionDisplay/View.php? I think the configuration of these two files belong together. Or what is the idea behind that?

File: src/Plugin/EntityBrowser/Widget/View.php (hunk 1 above)
Comment (boobaa) [inline src/Plugin/EntityBrowser/Widget/View.php:225]: Oh, thanks, this is the easiest, cleanest and leanest way of solving this. Updated the PR accordingly.

**Comment by slashrsm** [comment]
Merged rebased and squshed branch. Commit is: 620b4eb
